cmd/tealdbg/local.go

```diff
@@ -535,13 +535,14 @@ func (r *LocalRunner) RunAll() error {
 	start := time.Now()
 
 	ep := logic.NewEvalParams(txngroup, &r.proto, &transactions.SpecialAddresses{})
-	ep.Debugger = r.debugger
 
 	var last error
 	for i := range r.runs {
 		run := &r.runs[i]
-		r.debugger.SaveProgram(run.name, run.program, run.source, run.offsetToLine, run.states)
-
+		if r.debugger != nil {
```

Comment:
Changes related to C2C - `RunAll()` doesn't cooperate without this check.

Reply:
While the highlighted issue isn't central to the PR, I'm trying to use it as an opportunity to further my understanding.  The question that arises in my mind is:  Why `Run` doesn't delegate to `RunAll`?

It might be easier to discuss live.  I'm assuming I'm missing basic intuition.  

Details:
* If `Run` invokes `RunAll`, then the check for optional debugger (among other logic) lives in one place without duplication.
* As far as I can tell, `Run` is intended to be a test facility only.  It seems to me its visibility can be made private to more clearly signal its intended usage.  
* I began local changes to delegate `Run` to `RunAll` and paused because I think I'm missing basic intuition.  Hoping some folks can help me fill in the intuition:  In what cases does it make sense to generate `len(runs) > 1`, but only test the first index (e.g. `TestDebugCostPooling`)?

Reply:
If I recall right `Run` used to be a simplified version of `RunAll` used in tests only.

Reply:
@algochoi @algorandskiy I opened https://github.com/algorand/go-algorand/pull/3805 to eliminate duplicate code paths.  Can you take a look?

We can resolve the thread once #3805 is either merged into this PR or #3805 is closed.